### PR TITLE
KAFKA-13998: JoinGroupRequestData 'reason' can be too large

### PR DESCRIPTION
**Pull request by jnh5y**
This fix follows the pattern which is established in `AbstractCoordinator.java` of setting the request reason with the method `requestRejoin`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by jnh5y** [comment]
I tried to find a way to create a unit test for this change, but I wasn't able to do so quickly.  If someone has a suggestion for how to do that, I'm happy to follow through with it.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: @jnh5y If I understand correctly, you are saying that `exception.getMessage()` contained the stack trace. Am I right? I wonder why. Usually, the message is not suppose to contain it, no? It is a bit annoying because we are losing information without the message here.

nit: You pass two arguments but have only one `%s` in the pattern.

Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:482]: I guess it is due to this:
```
    public Throwable(Throwable cause) {
        fillInStackTrace();
        detailMessage = (cause==null ? null : cause.toString());
        this.cause = cause;
    }
``` 
`cause.toString()` contains the stack trace.

Comment (jnh5y) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Thanks for catching that I passed in two arguments!  (That would have prevented this PR from fixing the issue...)

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: nit: How about `rebalance failed due to %s`? `an exception` is redundant in my opinion.

**Review by dajac — COMMENTED** [review]
I have two general comments:
* Regarding the test, you might be able to get something based on https://github.com/apache/kafka/commit/ce7788aadaf989214c70e4df7ec286ed080e8cd1#diff-09c07df7cad69e6e7277725b13f69d1963fd843fe6283734334ace472171d3a4R2824.
* I wonder if we should limit the `reason` when we build `JoinGroupRequestData`. We could pick the first 255 chars for instance. That would prevent us from hitting this in the future. I am not sure what a good limit would be.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Would you have an example of the exception that was used to hit this?

Comment (jnh5y) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Unfortunately, I do not have an example of the exception.  I wish I did.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (jnh5y) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: +1.  Making that change.

**Comment by dajac** [comment]
@jnh5y Have you been able to come up with a unit test?

**Comment by jnh5y** [comment]
> @jnh5y Have you been able to come up with a unit test?

I have not yet; I'm traveling a little bit.  Don't let me hold up progress; I'll make an effort to get to it next week.

**Comment by ableegoldman** [comment]
Given that this has been actively causing problems for users (and PD alerts for us) would it be reasonable to merge this as-id and then let Jim follow up with a unit test when he's done traveling? WDYT @dajac ?

**Review by ableegoldman — APPROVED** [review]
LGTM -- I do like @dajac's suggestion to explicitly trim the reason to a reasonable length, but personally I would rather see this fix get merged ASAP and then have @jnh5y follow up with that fix along with the unit test

**Review by Kvicii — APPROVED** [review]

**Comment by dajac** [comment]
@ableegoldman Sure. Let me have another look at it on Monday and I will merge it.

**Comment by cadonna** [comment]
I do not see the need for such hurry. The build failures seem to be related to this PR. So those need to be fixed anyways. Code freeze for 3.3.0 is on July 20th, so plenty of time to add the length limitation and the unit test. 
If we can really not wait that @jnh5y is back we should open a new PR with all the stuff in place as soon as possible.

**Comment by cadonna** [comment]
> I wonder if we should limit the reason when we build JoinGroupRequestData. We could pick the first 255 chars for instance. That would prevent us from hitting this in the future. I am not sure what a good limit would be.

@dajac is there a specific reason you propose 255? Has this to do with the size of the entire request? If yes, on what does the size of the request depend on besides the reason? Is there a way to calculate how long the reason can be?

I am not familiar with the error this the too long reason caused, so bear with me, if those are stupid question.

**Comment by dajac** [comment]
> I do not see the need for such hurry. The build failures seem to be related to this PR. So those need to be fixed anyways. Code freeze for 3.3.0 is on July 20th, so plenty of time to add the length limitation and the unit test. If we can really not wait that @jnh5y is back we should open a new PR with all the stuff in place as soon as possible.

We should definitely fix the failing tests. The PR is not ready to be merged at the moment. I will quickly fix it so we can merge it.

**Comment by dajac** [comment]
> > I wonder if we should limit the reason when we build JoinGroupRequestData. We could pick the first 255 chars for instance. That would prevent us from hitting this in the future. I am not sure what a good limit would be.
> 
> @dajac is there a specific reason you propose 255? Has this to do with the size of the entire request? If yes, on what does the size of the request depend on besides the reason? Is there a way to calculate how long the reason can be?
> 
> I am not familiar with the error this the too long reason caused, so bear with me, if those are stupid question.

No specific reason for the proposed 255. My reasoning is that the reason should be rather small. We have actually two upper bounds: 1) the `reason` field's length is limited to `0x7fff`, and 2) the require size could also be limited on the broker side. I am actually surprised that we hit 1) here.

I think keeping the first 250 (or even 500) characters of the `reason` should be informative enough on the broker side and that would avoid being subject to hitting 1) in the future.

**Comment by cadonna** [comment]
> We should definitely fix the failing tests. The PR is not ready to be merged at the moment. I will quickly fix it so we can merge it.

Sounds good!

**Comment by cadonna** [comment]
@jnh5y I do not completely understand what you mean by:

> I tried to find a way to create a unit test for this change, but I wasn't able to do so quickly.

I have the impression that the unit test coverage is sufficient in `AbstractCoordinator#testRejoinReason`.

So I am fine with merging.

**Review by cadonna — APPROVED** [review]

**Comment by cadonna** [comment]
I am +1 for introducing a limit of 255 characters for the reason in a follow-up PR.

**Review by dajac — APPROVED** [review]
LGTM. I will merge it when the build successfully completes.

**Comment by dajac** [comment]
Merged to trunk and 3.2.

**Comment by dajac** [comment]
I filed a JIRA for the limit part: https://issues.apache.org/jira/browse/KAFKA-14013. @jnh5y I am happy to do this part if you don't have cycles for it.

**Comment by jnh5y** [comment]
> @jnh5y I do not completely understand what you mean by:
> 
> > I tried to find a way to create a unit test for this change, but I wasn't able to do so quickly.
> 
> I have the impression that the unit test coverage is sufficient in `AbstractCoordinator#testRejoinReason`.

Hi @cadonna, good question!  At the point when I said that, I hadn't seen `AbstractCoordinator#testRejoinReason`.

I was looking for a test which would call `AbstractCoordinator.joinGroupIfNeeded` and pass in an exception which was too large.  

I'm not still seeing how to modify that particular test to pass in an arbitrary exception.  That said, I'm happy with @dajac's solution to call `requestRejoin`.  In the follow-up ticket, we could test that this function works as we expect.

**Comment by jnh5y** [comment]
@dajac Thanks for fixing the PR and merging it!

If that ticket can wait a little bit, I'd like to find the time to do it.  I can find time before the July 20th code freeze.

**Comment by dajac** [comment]
> I'm not still seeing how to modify that particular test to pass in an arbitrary exception. That said, I'm happy with @dajac's solution to call requestRejoin. In the follow-up ticket, we could test that this function works as we expect.

@jnh5y Yeah, you cannot rely on providing the error in the response in your case. Instead, you have to find a place where you can actually throw a custom exception. That could be with a custom assignor for instance.

> If that ticket can wait a little bit, I'd like to find the time to do it. I can find time before the July 20th code freeze.

As long as we can get it in before the code freeze, I am fine.

**Comment by jnh5y** [comment]
> @jnh5y Yeah, you cannot rely on providing the error in the response in your case. Instead, you have to find a place where you can actually throw a custom exception. That could be with a custom assignor for instance.

I'm still learning the Kafka codebase, is there an example of using a custom assignor that is easy to understand?  Also, should we go to the trouble to test this particular code path or is it sufficient to verify that the `reason` is limited when the function setting it is called?  (I can see either argument.)

**Comment by dajac** [comment]
> > @jnh5y Yeah, you cannot rely on providing the error in the response in your case. Instead, you have to find a place where you can actually throw a custom exception. That could be with a custom assignor for instance.
> 
> I'm still learning the Kafka codebase, is there an example of using a custom assignor that is easy to understand? Also, should we go to the trouble to test this particular code path or is it sufficient to verify that the `reason` is limited when the function setting it is called? (I can see either argument.)

@jnh5y Personally, I am fine with the test that we have in this PR. https://issues.apache.org/jira/browse/KAFKA-14013 is the important missing part in my opinion.

**Comment by cadonna** [comment]
I agree with @dajac that KAFKA-14013 is much more urgent.
As far as I know @jnh5y is on vacation until the end of next week.

**Comment by jnh5y** [comment]
> I agree with @dajac that [KAFKA-14013](https://issues.apache.org/jira/browse/KAFKA-14013) is much more urgent. As far as I know @jnh5y is on vacation until the end of next week.

Yes, I'm traveling at the minute.  My plan is look at KAFKA-14013 Thursday / Friday of next week.  Since it should be quick, hopefully that works for everyone.

**Pull request merged**
